BatchCreate task doesn't work with ION files

According to the task documentation, this flow should work -- it uses the same ION file as in unit tests:

```yaml
id: weaviate_batch
namespace: dev

tasks:
  - id: extract
    type: io.kestra.plugin.fs.http.Download
    uri: https://gist.githubusercontent.com/anna-geller/cb04f310f12c96d58cc98f1aadfc8bb1/raw/2d77a6dccfd701d05a07684176dd37e0e2e6036e/ingest.ion

  - id: batch_load
    type: io.kestra.plugin.weaviate.BatchCreate
    host: "https://demo-oczq9ryw.weaviate.network"
    apiKey: xxx
    className: Titles
    objects: "{{ outputs.extract.uri }}"
```

This throws an error:

```
ERROR Illegal character in path at index 0: {{ outputs.extract.uri }}
2023-10-04T19:32:33.246Z TRACE java.lang.IllegalArgumentException: Illegal character in path at index 0: {{ outputs.extract.uri }}
	at java.base/java.net.URI.create(URI.java:911)
	at io.kestra.plugin.weaviate.BatchCreate.run(BatchCreate.java:86)
	at io.kestra.plugin.weaviate.BatchCreate.run(BatchCreate.java:27)
	at io.kestra.core.runners.Worker$WorkerThread.run(Worker.java:662)
Caused by: java.net.URISyntaxException: Illegal character in path at index 0: {{ outputs.extract.uri }}
	at java.base/java.net.URI$Parser.fail(URI.java:2963)
	at java.base/java.net.URI$Parser.checkChars(URI.java:3134)
	at java.base/java.net.URI$Parser.parseHierarchical(URI.java:3216)
	at java.base/java.net.URI$Parser.parse(URI.java:3175)
	at java.base/java.net.URI.<init>(URI.java:628)
	at java.base/java.net.URI.create(URI.java:909)
	... 3 more
```

There is no test or example for an array input — trying to pass both a list of strings or list of maps leads to error: `https, Failed ids: ea339990-aff2-40dd-9898-a852afeddfa8, f144f6fc-dbaa-455f-a907-62a6b51a222f`

### Environment Information

- Kestra Version: 0.13.0

--- a/src/main/java/io/kestra/plugin/weaviate/BatchCreate.java
+++ b/src/main/java/io/kestra/plugin/weaviate/BatchCreate.java
@@ -6,6 +6,7 @@ import io.kestra.core.models.annotations.PluginProperty;
 import io.kestra.core.models.tasks.RunnableTask;
 import io.kestra.core.runners.RunContext;
 import io.kestra.core.serializers.FileSerde;
+import io.kestra.core.utils.Rethrow;
 import io.reactivex.BackpressureStrategy;
 import io.reactivex.Flowable;
 import io.swagger.v3.oas.annotations.media.Schema;
@@ -24,6 +25,8 @@ import java.net.URI;
 import java.util.*;
 import java.util.stream.Collectors;
 
+import static io.kestra.core.utils.Rethrow.throwFunction;
+
 @SuperBuilder
 @ToString
 @Getter
@@ -41,9 +44,20 @@ import java.util.stream.Collectors;
                 "host: localhost:8080",
                 "apiKey: some_api_key",
                 "className: WeaviateObject",
-                "objects: " +
-                    "   textField: \"text\"",
-                "   numField: 32"
+                "objects: ",
+                "  - textField: \"text\"",
+                "    numField: 32",
+                "  - textField: \"another text\"",
+                "    numField: 64"
+            }
+        ),
+        @Example(
+            title = "Send batch object creation request to a Weaviate database using a ION input file resulting from another database query",
+            code = {
+                "host: localhost:8080",
+                "apiKey: some_api_key",
+                "className: WeaviateObject",
+                "objects: {{ outputs.sql.uri }}"
             }
         )
     }
@@ -75,15 +89,15 @@ public class BatchCreate extends WeaviateConnection implements RunnableTask<Batc
 
         if (objects instanceof List) {
             weaviateObjects = ((List<Map<String, Object>>) objects).stream()
-                .map(param -> WeaviateObject.builder()
+                .map(throwFunction(param -> WeaviateObject.builder()
                     .id(UUID.randomUUID().toString())
                     .className(className)
-                    .properties(param)
+                    .properties(runContext.render(param))
                     .build()
-                ).toList();
+                )).toList();
         } else if (objects instanceof String uri) {
             try (BufferedReader reader = new BufferedReader(new InputStreamReader(
-                runContext.uriToInputStream(URI.create(uri))
+                runContext.uriToInputStream(URI.create(runContext.render(uri)))
             ))) {
                 weaviateObjects = Flowable.create(FileSerde.reader(reader, Map.class), BackpressureStrategy.BUFFER)
                     .map(map -> WeaviateObject.builder()

--- a/src/main/java/io/kestra/plugin/weaviate/WeaviateConnection.java
+++ b/src/main/java/io/kestra/plugin/weaviate/WeaviateConnection.java
@@ -10,7 +10,8 @@ import io.weaviate.client.v1.auth.exception.AuthException;
 import lombok.*;
 import lombok.experimental.SuperBuilder;
 
-import javax.validation.constraints.NotBlank;
+import java.util.Collections;
+import java.util.Map;
 
 @SuperBuilder
 @ToString
@@ -18,18 +19,23 @@ import javax.validation.constraints.NotBlank;
 @Getter
 @NoArgsConstructor
 public abstract class WeaviateConnection extends Task implements WeaviateConnectionInterface {
+    private String url;
 
-    @Builder.Default
-    private String scheme = "https";
-
-    @NotBlank
-    private String host;
-
-    @NotBlank
     private String apiKey;
 
+    @Builder.Default
+    private Map<String, String> headers = Collections.emptyMap();
+
     protected WeaviateClient connect(RunContext context) throws AuthException, IllegalVariableEvaluationException {
-        Config config = new Config(context.render(scheme), context.render(host));
+        String renderedUrl = context.render(url);
+        int schemeSeparatorIdx = renderedUrl.indexOf("://");
+        String scheme = schemeSeparatorIdx == -1 ? "https" : renderedUrl.substring(0, schemeSeparatorIdx);
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        Config config = new Config(
+            scheme,
+            renderedUrl.substring(schemeSeparatorIdx == -1 ? 0 : schemeSeparatorIdx + 3),
+            context.render((Map) headers)
+        );
 
         if (apiKey == null) {
             return new WeaviateClient(config);

--- a/src/main/java/io/kestra/plugin/weaviate/WeaviateConnectionInterface.java
+++ b/src/main/java/io/kestra/plugin/weaviate/WeaviateConnectionInterface.java
@@ -4,28 +4,27 @@ import io.kestra.core.models.annotations.PluginProperty;
 import io.swagger.v3.oas.annotations.media.Schema;
 
 import javax.validation.constraints.NotBlank;
+import java.util.Map;
 
 public interface WeaviateConnectionInterface {
-
     @Schema(
-        title = "Connection scheme. Default is https"
-    )
-    @PluginProperty(dynamic = true)
-    String getScheme();
-
-    @Schema(
-        title = "Connection host",
-        description = "Example: localhost:8080"
+        title = "Connection url",
+        description = "Example: localhost:8080 or https://secured.weaviate:8080"
     )
     @NotBlank
     @PluginProperty(dynamic = true)
-    String getHost();
+    String getUrl();
 
     @Schema(
         title = "Connection api key",
         description = "If not provided, the anonymous authentication scheme will be used"
     )
-    @NotBlank
     @PluginProperty(dynamic = true)
     String getApiKey();
+
+    @Schema(
+        title = "Additional headers to add to the request"
+    )
+    @PluginProperty(dynamic = true)
+    Map<String, String> getHeaders();
 }

--- a/src/test/java/io/kestra/plugin/weaviate/BatchCreateTest.java
+++ b/src/test/java/io/kestra/plugin/weaviate/BatchCreateTest.java
@@ -26,13 +26,12 @@ public class BatchCreateTest extends WeaviateTest {
 
     @Test
     public void testBatchCreateWithParameters() throws Exception {
-        RunContext runContext = runContextFactory.of();
+        RunContext runContext = runContextFactory.of(Map.of("title", "test success"));
 
-        List<Map<String, Object>> objectsToCreate = List.of(Map.of("title", "test success"));
+        List<Map<String, Object>> objectsToCreate = List.of(Map.of("title", "{{title}}"));
 
         BatchCreate.Output batchOutput = BatchCreate.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .className(CLASS_NAME)
             .objects(objectsToCreate)
             .build()
@@ -41,12 +40,11 @@ public class BatchCreateTest extends WeaviateTest {
         assertThat(batchOutput.getCreatedCount(), is(1));
         assertThat(batchOutput.getUri(), notNullValue());
 
-        assertThat(readObjectsFromStream(runContext.uriToInputStream(batchOutput.getUri())), is(objectsToCreate));
+        assertThat(readObjectsFromStream(runContext.uriToInputStream(batchOutput.getUri())), is(List.of(Map.of("title", "test success"))));
     }
 
     @Test
     public void testBatchCreateWithUri() throws Exception {
-        RunContext runContext = runContextFactory.of();
 
         String fileName = "weaviate-objects.ion";
         URL resource = BatchCreate.class.getClassLoader().getResource(fileName);
@@ -56,11 +54,11 @@ public class BatchCreateTest extends WeaviateTest {
             new FileInputStream(Objects.requireNonNull(resource).getFile())
         );
 
+        RunContext runContext = runContextFactory.of(Map.of("uri", uri.toString()));
         BatchCreate.Output batchOutput = BatchCreate.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .className(CLASS_NAME)
-            .objects(uri.toString())
+            .objects("{{uri}}")
             .build()
             .run(runContext);
 

--- a/src/test/java/io/kestra/plugin/weaviate/DeleteTest.java
+++ b/src/test/java/io/kestra/plugin/weaviate/DeleteTest.java
@@ -11,7 +11,7 @@ import java.util.List;
 import java.util.Map;
 
 import static org.hamcrest.MatcherAssert.assertThat;
-import static org.hamcrest.Matchers.*;
+import static org.hamcrest.Matchers.is;
 
 public class DeleteTest extends WeaviateTest {
     @Inject
@@ -24,8 +24,7 @@ public class DeleteTest extends WeaviateTest {
         List<Map<String, Object>> parameters = List.of(Map.of("title", "test success"));
 
         BatchCreate.Output batchOutput = BatchCreate.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .className(CLASS_NAME)
             .objects(parameters)
             .build()
@@ -34,19 +33,18 @@ public class DeleteTest extends WeaviateTest {
         assertThat(batchOutput.getCreatedCount(), is(1));
 
         FetchOutput queryOutput = Query.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .query("""
-                   {
-                          Get {
-                            %s {
-                              _additional {
-                                id
-                              }
-                            }
-                          }
-                        }
-                   """.formatted(CLASS_NAME))
+                {
+                       Get {
+                         %s {
+                           _additional {
+                             id
+                           }
+                         }
+                       }
+                     }
+                """.formatted(CLASS_NAME))
             .fetchType(FetchType.FETCH_ONE)
             .build()
             .run(runContext);
@@ -56,8 +54,7 @@ public class DeleteTest extends WeaviateTest {
         String id = (String) ((Map<String, Object>) queryOutput.getRow().get("_additional")).get("id");
 
         Delete.Output deleteOutput = Delete.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .className(CLASS_NAME)
             .id(id)
             .build()
@@ -69,19 +66,18 @@ public class DeleteTest extends WeaviateTest {
         assertThat(deleteOutput.getDeletedCount(), is(1L));
 
         queryOutput = Query.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .query("""
-                      {
-                        Get {
-                            %s {
-                              _additional {
-                                id
-                              }
-                            }
-                        }
-                      }
-                   """.formatted(CLASS_NAME))
+                   {
+                     Get {
+                         %s {
+                           _additional {
+                             id
+                           }
+                         }
+                     }
+                   }
+                """.formatted(CLASS_NAME))
             .fetchType(FetchType.FETCH)
             .build()
             .run(runContext);
@@ -115,8 +111,7 @@ public class DeleteTest extends WeaviateTest {
         );
 
         BatchCreate.Output batchOutput = BatchCreate.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .className(CLASS_NAME)
             .objects(createdObjects)
             .build()
@@ -125,8 +120,7 @@ public class DeleteTest extends WeaviateTest {
         assertThat(batchOutput.getCreatedCount(), is(3));
 
         Delete.Output deleteOutput = Delete.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .className(CLASS_NAME)
             .filter(Map.of(
                 "title", "success",
@@ -143,19 +137,18 @@ public class DeleteTest extends WeaviateTest {
         assertThat(deleteOutput.getDeletedCount(), is(1L));
 
         FetchOutput queryOutput = Query.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .query("""
-                      {
-                        Get {
-                            %s {
-                              _additional {
-                                id
-                              }
-                            }
-                        }
-                      }
-                   """.formatted(CLASS_NAME))
+                   {
+                     Get {
+                         %s {
+                           _additional {
+                             id
+                           }
+                         }
+                     }
+                   }
+                """.formatted(CLASS_NAME))
             .fetchType(FetchType.FETCH)
             .build()
             .run(runContext);
@@ -163,8 +156,7 @@ public class DeleteTest extends WeaviateTest {
         assertThat(queryOutput.getSize(), is(2L));
 
         deleteOutput = Delete.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .className(CLASS_NAME)
             .filter(Map.of(
                 "title", "success"
@@ -177,19 +169,18 @@ public class DeleteTest extends WeaviateTest {
         assertThat(deleteOutput.getDeletedCount(), is(2L));
 
         queryOutput = Query.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .query("""
-                      {
-                        Get {
-                            %s {
-                              _additional {
-                                id
-                              }
-                            }
-                        }
-                      }
-                   """.formatted(CLASS_NAME))
+                   {
+                     Get {
+                         %s {
+                           _additional {
+                             id
+                           }
+                         }
+                     }
+                   }
+                """.formatted(CLASS_NAME))
             .fetchType(FetchType.FETCH)
             .build()
             .run(runContext);

--- a/src/test/java/io/kestra/plugin/weaviate/QueryTest.java
+++ b/src/test/java/io/kestra/plugin/weaviate/QueryTest.java
@@ -38,8 +38,7 @@ public class QueryTest extends WeaviateTest {
         List<Map<String, Object>> parameters = List.of(Map.of("title", "test success"));
 
         BatchCreate.Output batchOutput = BatchCreate.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .className(CLASS_NAME)
             .objects(parameters)
             .build()
@@ -48,8 +47,7 @@ public class QueryTest extends WeaviateTest {
         assertThat(batchOutput.getCreatedCount(), is(1));
 
         FetchOutput queryOutput = Query.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .query("""
                 {
                        Get {
@@ -86,8 +84,7 @@ public class QueryTest extends WeaviateTest {
         );
 
         BatchCreate.Output batchOutput = BatchCreate.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .className(CLASS_NAME)
             .objects(objectsToCreate)
             .build()
@@ -96,8 +93,7 @@ public class QueryTest extends WeaviateTest {
         assertThat(batchOutput.getCreatedCount(), is(2));
 
         FetchOutput queryOutput = Query.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .query(QUERY.formatted(CLASS_NAME))
             .fetchType(FetchType.FETCH)
             .build()
@@ -123,8 +119,7 @@ public class QueryTest extends WeaviateTest {
         );
 
         BatchCreate.Output batchOutput = BatchCreate.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .className(CLASS_NAME)
             .objects(objectsToCreate)
             .build()
@@ -133,8 +128,7 @@ public class QueryTest extends WeaviateTest {
         assertThat(batchOutput.getCreatedCount(), is(2));
 
         FetchOutput queryOutput = Query.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .query(QUERY.formatted(CLASS_NAME))
             .fetchType(FetchType.STORE)
             .build()

--- a/src/test/java/io/kestra/plugin/weaviate/SchemaCreateTest.java
+++ b/src/test/java/io/kestra/plugin/weaviate/SchemaCreateTest.java
@@ -26,8 +26,7 @@ public class SchemaCreateTest extends WeaviateTest {
         Map<String, List<String>> parameters = Map.of("title", List.of("text"));
 
         SchemaCreate.Output schemeOutput = SchemaCreate.builder()
-            .scheme(SCHEME)
-            .host(HOST)
+            .url(URL)
             .className(CLASS_NAME)
             .fields(parameters)
             .build()

--- a/src/test/java/io/kestra/plugin/weaviate/WeaviateTest.java
+++ b/src/test/java/io/kestra/plugin/weaviate/WeaviateTest.java
@@ -16,8 +16,8 @@ import java.util.Map;
 
 @MicronautTest
 public abstract class WeaviateTest {
-    protected static final String SCHEME = "http";
     protected static final String HOST = "localhost:51525";
+    protected static final String URL = "http://localhost:51525";
     protected static final String CLASS_NAME = "WeaviateTest";
 
     @AfterEach
@@ -26,7 +26,7 @@ public abstract class WeaviateTest {
     }
 
     protected WeaviateClient client() {
-        return new WeaviateClient(new Config(SCHEME, HOST));
+        return new WeaviateClient(new Config("http", HOST));
     }
 
     protected List<Map> readObjectsFromStream(InputStream inputStream) throws Exception {

